python/lsst/daf/butler/transfers/_context.py

```diff
@@ -207,13 +238,106 @@ def _finish(self) -> None:
+            for run in datasetsByRun.keys():
+                self._collections[run] = self._registry._collections.find(run)
+        for collectionName in self._computeSortedCollections():
+            self._backend.saveCollection(self._collections[collectionName])
+        # Continuing for DM-26324: then we need to either make DatasetType
+        # sortable directly or sort the iteration below by its name (as well as
+        # run).
+        for datasetType in self._datasets.keys():
+            for run in self._datasets[datasetType].keys():
+                # Again, for DM-26324: And after that, that we need to sort the
+                # FileDataset objects in the third argument below (maybe by
+                # filename?) and the lists of DatasetRef within those (I'd use
+                # the aforementioned new DataCoordinate sort method, because
+                # I'm not sure dataset_id values are going to be reliably
+                # deterministic themselves).
+                records = self._datasets[datasetType][run]
+                self._backend.saveDatasets(datasetType, run, *records)
+        # Export associations between datasets and collections.  These need to
+        # be sorted (at two levels; they're dicts) or created more
+        # deterministically, too, which probably involves more data ID sorting.
+        for collection, associations in self._computeDatasetAssociations().items():
+            self._backend.saveDatasetAssociations(collection, self._collections[collection].type,
+                                                  associations)
         self._backend.finish()
+
+    def _computeSortedCollections(self) -> List[str]:
+        """Sort collections in a way that is both deterministic and safe
+        for registering them in a new repo in the presence of nested chains.
+
+        This method is intended for internal use by `RepoExportContext` only.
+
+        Returns
+        -------
+        names: `List` [ `str` ]
+            Ordered list of collection names.
+        """
+        # Split collections into CHAINED and everything else, and just
+        # sort "everything else" lexicographically since there are no
+        # dependencies.
+        chains: Dict[str, List[str]] = {}
+        result: List[str] = []
+        for record in self._collections.values():
+            if record.type is CollectionType.CHAINED:
+                assert isinstance(record, ChainedCollectionRecord)
+                chains[record.name] = [child for child, _ in record.children]
+            else:
+                result.append(record.name)
+        result.sort()
+        # Sort all chains topologically, breaking ties lexicographically.
+        # Append these to 'result' and remove them from 'chains' as we go.
+        while chains:
+            unblocked = {
+                parent for parent, children in chains.items()
+                if not any(child in chains.keys() for child in children)
```

Comment:
is the keys() needed here? Normally `__contains__` would work on the dict directly.

Reply:
Not needed, I just thought it was easier to read.  I know `.keys` is basically always visual noise for some people, but when it's not extremely obvious that the target of `__contains__` or `__iter__` is a mapping I find it very helpful to have it called out.

Reply:
Ok, it looks odd to me because I'm then wondering what's special about the `keys` that means that `in` doesn't work (and surely it makes python work harder since it then has to generate keys and do the `__contains__` on that?)